crates/oxc_semantic/src/builder.rs

```diff
@@ -674,6 +667,32 @@ impl<'a> Visit<'a> for SemanticBuilder<'a> {
         self.leave_node(kind);
     }
 
+    fn visit_block_statement(&mut self, it: &BlockStatement<'a>) {
+        let kind = AstKind::BlockStatement(self.alloc(it));
+        self.enter_node(kind);
+
+        // Move all bindings from parent scope to current scope
+        // to make it easier to resole references and check redeclare errors.
+        // NOTE: enter_scope hasn't been called yet, so `current_scope_id`` is the parent scope of the block's scope.
+        let is_catch_clause = self.current_scope_flags().is_catch_clause();
+        let parent_bindings = if is_catch_clause {
+            Some(self.scope.get_bindings_mut(self.current_scope_id).drain(..).collect::<Bindings>())
+        } else {
+            None
+        };
+
+        self.enter_scope(ScopeFlags::empty(), &it.scope_id);
+
+        if let Some(parent_bindings) = parent_bindings {
+            *self.scope.get_bindings_mut(self.current_scope_id) = parent_bindings;
+        }
```

Comment:
I suggest:

```suggestion
        let parent_scope_id = self.current_scope_id;

        self.enter_scope(ScopeFlags::empty(), &it.scope_id);

        // Move all bindings from parent scope to current scope
        // to make it easier to resole references and check redeclare errors.
        // NOTE: enter_scope hasn't been called yet, so `current_scope_id`` is the parent scope of the block's scope.
        if self.scope.get_flags(parent_scope_id).is_catch_clause() {
            let parent_bindings = std::mem::take(self.scope.get_bindings_mut(parent_scope_id));
            *self.scope.get_bindings_mut(self.current_scope_id) = parent_bindings;
        }
```

1 less branch, and a bit simpler.

NB `take` is better than `drain` + `collect` I think as `drain` + `collect` empties the original hash map, allocates a new one, and copies the contents over. `take` just grabs it and moves it.

Reply:
Thanks!